ee/kernel/src/timer_alarm.c

```diff
@@ -0,0 +1,264 @@
+
+static int t2Overflow(int ca) {
+    (void)ca;
+
+    // Check if we actually triggered an alarm
+    if (event_queue) {
+        if (ps2_clock() > (event_queue->scheduled_time >> 8)) {
+            // Remove the top of the queue
+            struct timer_alarm_t *gone = event_queue;
+            event_queue = event_queue->next;
+            if (event_queue)
+                event_queue->prev = NULL;
+            gone->prev = NULL;
+            gone->next = NULL;
+            gone->scheduled_time = 0;
+            Timer2Resched(event_queue);
+
+            if (gone->callback)
+                gone->callback(gone, gone->usr_arg);
+        }
+        else  // Timer might require several IRQs (for timers over ~100ms)
+            Timer2Resched(event_queue);
+    }
+
+    // Clear both compare and overflow flags (since we can use either)
+    *T2_MODE |= (3 << 10);
+
+    ExitHandler();
+    return -1;
+}
+
+/** Initializes a timer alarm struct so that it can be used.
+ *
+ * @param alarm Alarm object to initialize.
+ *
+ * Initializes the alarm object to be used. The object must be kept
+ * around as long as it is being used.
+ */
+void InitializeTimerAlarm(struct timer_alarm_t *alarm) {
+    // Just return an empty struct, that's not in the event queue
+    alarm->prev = NULL;
+    alarm->next = NULL;
+    alarm->scheduled_time = 0;
+    alarm->timer_cycles = 0;
+    alarm->callback = NULL;
+    alarm->usr_arg = NULL;
+}
+
+/** Same as StopTimerAlarm but for IRQ handlers
+ *
+ */
+void iStopTimerAlarm(struct timer_alarm_t *alarm) {
+    // Bail if the alarm is not set.
+    if (!alarm->scheduled_time)
+        return;
+
+    // Check if the event is the next event in the queue.
+    if (event_queue == alarm) {
+        // Remove it from the front of the queue
+        if (alarm->next)
+            alarm->next->prev = NULL;
+        event_queue = alarm->next;
+
+        // Reschedule next event
+        Timer2Resched(event_queue);
+    }
+    else {
+        // Simply remove it, since it is not yet scheduled.
+        alarm->prev->next = alarm->next;
+        if (alarm->next)
+            alarm->next->prev = alarm->prev;
+    }
+
+    // Cleanup the structure
+    alarm->next = NULL;
+    alarm->prev = NULL;
+    alarm->scheduled_time = 0;
+}
+
+/** Stop an alarm
+ *
+ * @param alarm Alarm object to be stopped.
+ *
+ * Stops the alarm. If the alarm was not armed it does nothing.
+ * When an alarm triggers it is automatically disabled, so user must
+ * call iStartTimerAlarm to re-enable it.
+ */
+void StopTimerAlarm(struct timer_alarm_t *alarm) {
+    u32 oldintr = DIntr();
+    iStopTimerAlarm(alarm);
+    if (oldintr)
+        EIntr();
+}
+
+/** Same as StartTimerAlarm but for IRQ handlers.
+ *
+ */
+void iStartTimerAlarm(struct timer_alarm_t *alarm) {
+    // Bail if the alarm is already set!
+    if (alarm->scheduled_time)
+        return;
+
+    // Calculate clock scheduling time
+    u64 sched_time = ps2_clock() * 256 + alarm->timer_cycles;
```

Comment:
`ps2_clock` belong to `ee/libc/src/time.c` which is initialised with `_ps2sdk_time_init`, here we can have something wrong if `_ps2sdk_time_init` is not called (it is currently a `weak function`).
Maybe we should move that implementation to `kernel` and do the init always?
Pinging @rickgaiser